File: SexyBiscuit.Engine/Gameplay/PlayerState.cs
using SexyBiscuit.Engine.Core;
using SexyBiscuit.Engine.Networking;

namespace SexyBiscuit.Engine.Gameplay;

/// <summary>
/// Per-player state that every machine in a session needs to see — name, score, ping.
/// Modelled on Unreal's <c>APlayerState</c>.
/// </summary>
/// <remarks>
/// Lives separately from the pawn so it survives death and respawn, and separately from the
/// controller so remote clients can read it for players they do not own. Mark fields you add
/// with <c>[Replicated]</c> to have the replication system carry them.
///
/// Every field below is replicated. None of them were, so a client that joined saw an empty
/// scoreboard; a game cannot fix that by subclassing, because shadowing a property would send it
/// twice under two names while <c>GameState</c> keeps reading the base.
/// </remarks>
public class PlayerState : Actor
{
    /// <summary>Display name shown in scoreboards and nameplates.</summary>
    [Replicated]
    public string PlayerName { get; set; } = "Player";

    /// <summary>Stable identifier for this player within the session.</summary>
    [Replicated]
    public int PlayerId { get; set; }

    /// <summary>Score in whatever unit the game mode uses.</summary>
    [Replicated]
    public int Score { get; set; }

    /// <summary>Team index; -1 means unassigned or free-for-all.</summary>
    [Replicated]
    public int TeamId { get; set; } = -1;

    /// <summary>Round-trip latency in milliseconds. Zero for the local player on a listen server.</summary>
    [Replicated]
    public int PingMs { get; set; }

    /// <summary>True when this state belongs to a bot rather than a human.</summary>
    [Replicated]
    public bool IsBot { get; set; }

    /// <summary>Raised whenever <see cref="Score"/> changes through <see cref="AddScore"/>.</summary>
    public SBEvent<int> ScoreChanged { get; } = new();

    public PlayerState() : base("PlayerState") { }

    /// <summary>Adds to the score and broadcasts <see cref="ScoreChanged"/> with the new total.</summary>
    public void AddScore(int delta)
    {
        Score += delta;
        ScoreChanged.Broadcast(Score);
    }
}

File: html5/src/gameplay/PlayerState.js
// -----------------------------------------------------------------------------
// PlayerState — one player's persistent data, surviving respawns.
// -----------------------------------------------------------------------------

import { Actor } from '../core/Actor.js';
import { registerActor } from '../core/TypeRegistry.js';
import { PropertyType as P } from '../core/PropertyTypes.js';
import { SBEvent } from '../core/SBEvent.js';

/** Score, name and team for one player. */
export class PlayerState extends Actor {
    static schema = {
        playerName: { type: P.String, default: 'Player' },
        teamId:     { type: P.Int, default: -1 },
        isBot:      { type: P.Bool, default: false },
    };

    constructor(name = 'PlayerState') {
        super(name);
        this.playerName = 'Player';
        this.playerId = 0;
        this.score = 0;
        this.teamId = -1;
        this.pingMs = 0;
        this.isBot = false;

        this.scoreChanged = new SBEvent();
    }

    /** Adds to the score and broadcasts the new total. */
    addScore(points) {
        this.score += points;
        this.scoreChanged.broadcast(this.score);
        return this.score;
    }
}
registerActor(PlayerState, { category: 'Gameplay', summary: 'One player’s persistent data.' });
